lib/streamlit/connections/snowflake_connection.py

```diff
@@ -0,0 +1,260 @@
+            _validate_connection_params(conn_kwargs)
+            return snowflake.connector.connect(**conn_kwargs)
+
+        if hasattr(snowflake.connector.connection, "CONFIG_MANAGER"):
+            # TODO(vdonato): Double-check that we want to map the "snowflake" connection
+            # name in a call to `st.connection` to the "default" connection in the
+            # Snowflake config file. This is aligned with the default behavior for the
+            # underlying python connector and seems more sane than having a "snowflake"
+            # connection in a snowflake-specific config file.
+            connection_name = (
+                "default"
+                if self._connection_name == "snowflake"
+                else self._connection_name
+            )
+            return snowflake.connector.connect(
+                connection_name=connection_name,
+                **kwargs,
+            )
+
+        # If we have a version of the Snowflake Python Connector installed that predates
+        # the ConfigManager, look for credentials in  ~/.snowsql/config as a last
+        # resort.
+        snowsql_config = load_from_snowsql_config_file(self._connection_name)
+        conn_kwargs = {**snowsql_config, **kwargs}
+        _validate_connection_params(conn_kwargs)
+
+        return snowflake.connector.connect(**conn_kwargs)
+
+    def query(
+        self,
+        sql: str,
+        *,  # keyword-only arguments:
+        ttl: float | int | timedelta | None = None,
+        show_spinner: bool | str = "Running `snowflake.query(...)`.",
+        params=None,
+    ) -> pd.DataFrame:
+        """Run a read-only SQL query.
+
+        This method implements both query result caching (with caching behavior
+        identical to that of using ``@st.cache_data``) as well as simple error handling/retries.
+
+        .. note::
+            Queries that are run without a specified ttl are cached indefinitely.
+
+        Parameters
+        ----------
+        sql : str
+            The read-only SQL query to execute.
+        ttl : float, int, timedelta or None
+            The maximum number of seconds to keep results in the cache, or
+            None if cached results should not expire. The default is None.
+        show_spinner : boolean or string
+            Enable the spinner. The default is to show a spinner when there is a
+            "cache miss" and the cached resource is being created. If a string, the value
+            of the show_spinner param will be used for the spinner text.
+        params : list, tuple, dict or None
+            List of parameters to pass to the execute method. The syntax used to pass
+            parameters is database driver dependent. Check your database driver
+            documentation for which of the five syntax styles, described in `PEP 249
+            paramstyle <https://peps.python.org/pep-0249/#paramstyle>`_, is supported.
+            Default is None.
```

Comment:
Need to update this with the snowflake specific docs and behavior. This might be the best cite: https://docs.snowflake.com/en/developer-guide/python-connector/python-connector-example#label-python-connector-binding-data

Reply:
Probably worth noting the default / recommended style and including an example somewhere too

Seems like both the client side approaches work on local

```python
conn = st.connection("snowflake")
# Broken, see comment below
# df = conn.query("SELECT %(col1)s, %(col2)s", ttl=0, params={'col1': 789, 'col2': 'test string3'})
df = conn.cursor().execute("SELECT %s, %s", (789, 'test string3')).fetch_pandas_all()
df2 = conn.cursor().execute("SELECT %(col1)s, %(col2)s", {'col1': 123, 'col2': 'foo bar'}).fetch_pandas_all()
st.dataframe(df)
st.dataframe(df2)
```

qmark doesn't seem to work by default (which I think matches the documentation). Guessing that's OK for now.

Reply:
ack -- will punt this to a followup PR along with other potential tweaks to docstrings